pingcap/tidb-tools
# mydumper_uploader: add a new tool, mydumper_uploader.

### What problem does this PR solve? 
Add a tool can upload the data exported by `mydumper` to AWS S3, and support breakpoints.

### What is changed and how it works?
Observe the following scheme: 
https://docs.google.com/document/d/1IbqyCIYue8enzy_SeLFCm2636gGxoWITleaJ_G1ZClQ/edit

### Check List 

Tests 

 - Unit test
[x] Slicer, which splits the files into slices. https://github.com/pingcap/tidb-tools/pull/156/files?file-filters%5B%5D=.go#diff-0edcdeef3d4b72b22603c924e2ba4bb0R65
[x] Checkpoint. https://github.com/pingcap/tidb-tools/pull/156/files?file-filters%5B%5D=.go#diff-a0a0151de42659eae526b9c690621580R29
 - Integration test
[x] Integration test using a mock file driver.  
https://github.com/pingcap/tidb-tools/pull/156/files?file-filters%5B%5D=.go#diff-5decd2a0e77d76cc740c3ad4fcbb8619R32
https://github.com/pingcap/tidb-tools/pull/156/files?file-filters%5B%5D=.go#diff-f375e90c9f534692be8a958294bec1e9R38
 - Manual test (add detailed scripts or steps below)
[x] AWSS3ServerDriver.https://github.com/pingcap/tidb-tools/pull/156/files?file-filters%5B%5D=.go#diff-5decd2a0e77d76cc740c3ad4fcbb8619R110
Fill in the relevant parameters and run test.
[x] Integration test using a real AWS account.
1. upload data to AWS S3 using this tool.
2. kill the tool process.
3. restore the upload process.
4. after uploaded, download the file in AWS S3.
5. check the integrity of files.

Code changes

 - Has exported function/method change
`FileUploader` https://github.com/pingcap/tidb-tools/pull/156/files?file-filters%5B%5D=.go#diff-7ed52f284eec9e72b6d4a67eb63644e4R30

`AWSS3FileUploaderDriver` https://github.com/pingcap/tidb-tools/pull/156/files?file-filters%5B%5D=.go#diff-839e971c93fbbce27d9b80392049e0d7R101

 - Has interface methods change

`FileUploaderDriver` https://github.com/pingcap/tidb-tools/pull/156/files?file-filters%5B%5D=.go#diff-839e971c93fbbce27d9b80392049e0d7R46
Related changes

 - Updated the documentation :`README.md`,`mydumper_uploader/README.md`
